### ISSUE TITLE
functional testing for parsing the config file

### ISSUE THREAD

**Issue by toogley** [neomutt/neomutt#1306]
Hey.

i think #1301 shows, that testing whether the compiled neomutt binary can parse a muttrc containing lots of different hooks (with different patterns) would be useful. That means, these testing only verify that neomutt can somehow parse the muttrc file without errors - but it doesn't verify that the functionality of these hooks/patterns is working correctly.

However, testing the latter is much more work, i think. Testing only the parsing is very very easy:

**muttrc:**
```
message-hook ~A 'set pager=builtin'
message-hook '~f freshmeat-news' 'set pager="less \"+/^  subject: .*\""'
```

```
[[ -z $(./neomutt -F test/muttrc -n -D  2>&1 1>/dev/null ) ]]  && echo "success" || echo "failed" 
```

How should we integrate that into our test suite? 
* At least at the moment, we cannot easily use `acutest.h`, because the config parsing is not independent. That means, running the above command is much easier than using acutest.h
* acutest provides a nice way to display which tests have failed  - i think it would not be a good idea to develop our own logic for these functional tests.

```
Test test_base64_encode...                      [   OK   ]
Test test_base64_decode...                      [   OK   ]
Test test_base64_lengths...                     [   OK   ]
Test test_rfc2047...                            [   OK   ]
Test test_md5...                                [   OK   ]
Test test_md5_ctx...                            [   OK   ]
Test test_md5_ctx_bytes...                      [   OK   ]
Test test_string_strfcpy...                     [   OK   ]
Test test_string_strnfcpy...                    [   OK   ]
SUCCESS: All unit tests have passed.
```

Opinions?

**Comment by toogley** [comment]
hm, i guess making the config parsing independent/easy testable by acutest, is the best solution?

 initially, i doubt that, because we probably would need a way to run functional test anyways - but these are at the moment far in the future.

**Comment by flatcap** [comment]
Hi @toogley,

tl;dr It can't be done yet; something would have to be sacrificed.

I did a little experiment, last night, to see what we could do.
There are two types of config to parse: variables and commands.
There are two types of command validation: syntax and meaning.

The new config code has made the variable parsing much clearer, but it still has dependencies.

- `parse_set()` deals with the `set x = y`, but it needs:
  - `myvar_{get,set,del}()` -- these should be moved into a separate file
  - `mutt_extract_token()` which is quite simple until you want to use \`s (backticks), then you need
    - 3 functions in filter.c, which isn't too bad apart from a reference to:
      - `MuttIndexWindow`!
  - `mutt_expand_path()` is tricky.  it relies on several config variables and then delegates to the mailbox code
    - `mx_is_imap()`, `mx_is_notmuch()`, etc
    - `mutt_alias_lookup()`
    - `mutt_default_save()` (hooks)

The command processing is similar.
We can check the _syntax_ of the command, but we require a lot of code to check the _meaning_.

**Comment by flatcap** [comment]
Unfortunately, the patterns are different, but the same.

The patterns aren't parsed until they're needed.
When it's needed, the pattern flags are looked up in a table and relevant data is found for comparison.

Some of these checks are easy to fake, others will require a Context (open mailbox), Header (open email), Envelope, Body, etc.

So, we **could** test the patterns, but we'd need quite a bit of test data and setup to make it happen.

There are [47 flags](https://github.com/neomutt/neomutt/blob/master/pattern.c#L1117-L1168) to test!
